idaes/apps/grid_integration/multiperiod/tests/test_price_taker.py

```diff
@@ -0,0 +1,1376 @@
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+
+from pathlib import Path
+import pytest
+import numpy as np
+import pandas as pd
+
+from pyomo.environ import (
+    Var,
+    Binary,
+    NonNegativeReals,
+    Constraint,
+    Expression,
+)
+
+import pyomo.environ as aml
+from pyomo.common.config import ConfigValue, In
+
+from idaes.core import FlowsheetBlock
+
+from idaes.core.base.process_base import declare_process_block_class
+from idaes.models.unit_models import SkeletonUnitModelData
+
+from idaes.apps.grid_integration import DesignModel, OperationModel
+
+import idaes.logger as idaeslog
+
+from idaes.apps.grid_integration.multiperiod.price_taker_model import PriceTakerModel
+
+import matplotlib.pyplot as plt
+
+
+@pytest.fixture
+def excel_data():
+    # DATA_DIR = Path(__file__).parent
+    # file_path = DATA_DIR / "FLECCS.xlsx"
+    # data = pd.read_excel(file_path, sheet_name=1)
+    DATA_DIR = Path(__file__).parent
+    file_path = DATA_DIR / "FLECCS_princeton.csv"
+    data = pd.read_csv(file_path)
+    return data
+
+
+@pytest.mark.unit
+def test_seed_value():
+    m = PriceTakerModel()
+
+    m.seed = 50
+
+    assert m.seed == 50
+
+
+@pytest.mark.unit
+def test_daily_data_size(excel_data):
+    m = PriceTakerModel()
+
+    # Generate price data for each hour of every day in the data
+    daily_data = m.generate_daily_data(excel_data["BaseCaseTax"])
+
+    # Check that there is a row for each horizon length in a representative day
+    assert len(daily_data) == m.horizon_length
+
+
+@pytest.mark.unit
+def test_determine_optimal_num_clusters(excel_data):
+    # Added a range for optimal cluster values based on how the
+    # plot appears visually. Test can be removed in the future if
+    # failure occurs. This may depend on scikit-learn and kneed and
+    # the interaction thereof.
+
+    # Older versions get n_clusters = 10, Newer versions n_clusters = 11
+    m = PriceTakerModel()
+
+    daily_data = m.generate_daily_data(excel_data["BaseCaseTax"])
+    n_clusters, inertia_values = m.get_optimal_n_clusters(daily_data)
+
+    assert 9 <= n_clusters <= 15
+
+
+@pytest.mark.unit
+def test_elbow_plot(excel_data):
+    m = PriceTakerModel()
+
+    daily_data = m.generate_daily_data(excel_data["BaseCaseTax"])
+    m.get_optimal_n_clusters(daily_data, plot=True)
+    test_fig = plt.gcf()
+    plt.close("all")
+
+    assert test_fig is not None
```

Comment:
Does this really test much of the actual code? It runs the plotting code, but all this assert does is make sure the resulting plot got closed which doesn't really test the plot. Note that testing plotting code is generally very hard - we generally only test the data used to generate the plot and leave the actual plotting untested.

Reply:
You're correct this just tests if the plot was successfully called; it does not check if the "correct" plot or "correct" data was displayed.

I'm inclined to _remove_ the plotting test. Please see my other review comment for further information on the optimal clusters code snippet ([link](https://github.com/IDAES/idaes-pse/pull/1358/files#r1591054731)).